etc/afpd/afpstats.c

```diff
@@ -56,34 +99,75 @@ static gpointer afpstats_thread(gpointer _data _U_)
+                                            G_DBUS_PROXY_FLAGS_NONE,
+                                            NULL, /* GDBusInterfaceInfo */
+                                            "org.freedesktop.DBus",
+                                            "/org/freedesktop/DBus",
+                                            "org.freedesktop.DBus",
+                                            NULL, /* GCancellable */
+                                            &error))) {
         LOG(log_error, logtype_afpd,"Couldn't create bus proxy");
         return NULL;
     }
 
-    if (!dbus_g_proxy_call(bus_proxy, "RequestName", &error,
-                           G_TYPE_STRING, "org.netatalk.AFPStats",
-                           G_TYPE_UINT, 0,
-                           G_TYPE_INVALID,
-                           G_TYPE_UINT, &request_name_result,
-                           G_TYPE_INVALID)) {
+    if (!g_dbus_proxy_call_sync(bus_proxy,
+                                "RequestName",
+                                g_variant_new("(as)", "org.netatalk.AFPStats", 0),
+                                G_DBUS_CALL_FLAGS_NONE,
+                                -1, /* timeout */
+                                NULL,
+                                &error)) {
         LOG(log_error, logtype_afpd, "Failed to acquire DBUS name: %s", error->message);
         return NULL;
     }
+#endif
+
+    static const gchar introspection_xml[] =
+        "<node>"
+        "  <interface name='org.netatalk.AFPStats'>"
+        "    <method name='GetUsers'>"
+        "      <arg name='ret' type='as' direction='out'/>"
+        "    </method>"
+        "  </interface>"
+        "</node>";
+
+    introspection_data = g_dbus_node_info_new_for_xml (introspection_xml, NULL);
+    g_assert (introspection_data != NULL);
 
     AFPStatsObj *obj = g_object_new(AFPSTATS_TYPE_OBJECT, NULL);
-    dbus_g_connection_register_g_object(bus, "/org/netatalk/AFPStats", G_OBJECT(obj));
+
+    registration_id = g_dbus_connection_register_object(bus,
+                                      "/org/netatalk/AFPStats",
+                                      introspection_data->interfaces[0],
+                                      &interface_vtable,
+                                      G_OBJECT(obj),
+                                      NULL,
+                                      &error);
+    g_assert (registration_id > 0);
+
+    request_name_result = g_bus_own_name(G_BUS_TYPE_SYSTEM,
+                                         "org.freedesktop.DBus",
+                                         G_DBUS_PROXY_FLAGS_NONE,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL);
 
     g_main_loop_run(thread_loop);
+
+    g_bus_unown_name(request_name_result);
+    g_dbus_node_info_unref(introspection_data);
```

Comment:
Explicitly undo the registration when you no longer want that object to be available on the bus:

```suggestion
    g_dbus_connection_unregister_object (bus, registration_id);
    g_dbus_node_info_unref(introspection_data);
```

You'll probably also want to free the `AFPStatsObj` with `g_object_unref`?

Reply:
Good call. Pushing a commit that frees `obj`.